rasterio/_features.pyx

```diff
@@ -338,13 +338,19 @@ def _rasterize(shapes, image, transform, all_touched, merge_alg):
             num_geoms * sizeof(OGRGeometryH))
         pixel_values = <double *>CPLMalloc(num_geoms * sizeof(double))
 
+        # initialize all geoms to NULL
```

Comment:
This was what fixed the core dumping. My theory is that when the exception is raised below, `geoms[i]` doesn't get set to anything. Without being initialized to NULL, the code in the `finally` statement attempts to free the memory that has not been allocated and a core dump occurs.